### PR TITLE
ENH: Add GroupL1Norm, closes #619

### LINKED ISSUE #619: "Isotropic" L1 norm on product spaces
We need to add a way to handle isotropic L1 norms through "group l1" norms. ProxImaL does this in a rather nice way by adding a separate [group_norm1](https://github.com/comp-imaging/ProxImaL/blob/master/proximal/prox_fns/group_norm1.py) class. I suggest we do something similar, probably calling it `GroupL1Norm`.

See #564 for overall issue. This is also needed for PR #587. This is also a prerequisite for #597.

### PR DESCRIPTION
**Pull request by adler-j**
Initial commit for `GroupL1Norm` needs #555 to proceed.

TODO:

- [ ] Add tests
- [ ] (optional) Add for the case where the inner exponent is infinity.
- [ ] Add some more doc to `IndicatorGroupL1UnitBall`

### CODE REVIEW

File: odl/solvers/functional/default_functionals.py [hunk 1]
```diff
@@ -118,6 +121,114 @@ def convex_conj(self):
...
+
+    @property
+    def proximal(self):
+        """Return the proximal factory of the functional.
+
+        See Also
+        --------
+        proximal_l1 : proximal factory for the L1-norm.
+        """
+        return proximal_l1(space=self.domain, isotropic=True)
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: Not correct either. You rather get in each point the proximal of the p-norm, which is the identity minus the projection onto the p'-ball. That projection is not computable in closed form except for p = 1, 2 or infty. We probably don't need arbitrary p now, but at least the well-known cases should be correct now.

File: odl/solvers/functional/default_functionals.py [hunk 2]
```diff
@@ -118,6 +121,114 @@ def convex_conj(self):
...
+        \| x \|_{\\times, p} =
+        \\sum_{i = 1}^n \\left( \\sum_{j = 1}^m x_{i,j}^p \\right)^{1/p}.
+
+    If the functional is defined on an :math:`L_p^m`-like space, the group
+    :math:`L_1`-norm is defined as
+
+    .. math::
+
+        \| x \|_{\\times, p} =
+        \\int_{\Omega} \\left(\\sum_{j = 1}^m x_{j}^p\\right)^{1/p} dx.
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: And by the way, why do you use `\times`? Why not `\|x\|_{1,2}`, indicating which norm in which "axis"?

**Comment by adler-j** [comment]
Too much stuff to fix before 0.5, I'll leave this one out for now.

**Comment by adler-j** [comment]
Fixed most if not all comments, added the convex conjugate and added them to the largescale tests, which succeed. New review please.

File: odl/solvers/functional/default_functionals.py [hunk 3]
```diff
@@ -124,6 +129,248 @@ def convex_conj(self):
...
+        else:
+            return 0
+
+    @property
+    def proximal(self):
+        """Return the proximal factory of the functional.
+
+        See Also
+        --------
+        proximal_l1 : proximal factory for the L1-norm.
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: wrong

File: odl/test/largescale/solvers/nonsmooth/proximal_operator_slow_test.py [hunk 4]
```diff
@@ -97,6 +99,9 @@ def functional(request, offset, dual, stepsize):
         func = func.translated(g)
 
     if dual:
+        if name == 'groupl11':
+            pytest.xfail('group l1-1 norm has no proximal')
```
Comment (adler-j) [inline odl/test/largescale/solvers/nonsmooth/proximal_operator_slow_test.py]: This is obviously a todo

File: odl/solvers/functional/default_functionals.py [hunk 5]
```diff
@@ -124,6 +129,248 @@ def convex_conj(self):
...
+                    divisor = np.power(functional.pointwise_norm(x), p - 1)
+                    np.divide(dividend, divisor, out=divisor,
+                              where=divisor != 0)
+                    return divisor
+
+        return GroupL1Gradient()
+
+    @property
+    def proximal(self):
+        """Return the proximal factory of the functional.
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]:
```
`proximal factory`
```

File: odl/solvers/functional/default_functionals.py [hunk 6]
```diff
@@ -124,6 +129,248 @@ def convex_conj(self):
...
+        Parameters
+        ----------
+        vfspace : `ProductSpace`
+            Space of vector fields on which the operator acts.
+            It has to be a product space of identical spaces, i.e. a
+            power space.
+        exponent : non-zero float, optional
+            Exponent of the norm in each point. Values between
+            0 and 1 are currently not supported due to numerical
+            instability.
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: infinity?

File: odl/solvers/functional/default_functionals.py [hunk 7]
```diff
@@ -124,6 +129,248 @@ def convex_conj(self):
...
+        --------
+        >>> space = odl.rn(2)
+        >>> pspace = odl.ProductSpace(space, 2)
+        >>> op = IndicatorGroupL1UnitBall(pspace)
+        >>> op([[0.1, 0.5], [0.2, 0.3]])
+        0
+        >>> op([[3, 3], [4, 4]])
+        inf
+
+        Can also change exponent of inner  (p) norm:
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: see above

File: odl/solvers/functional/default_functionals.py (hunk 3 above)
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: see above

File: odl/solvers/functional/default_functionals.py [hunk 8]
```diff
@@ -124,6 +129,248 @@ def convex_conj(self):
...
+            return proximal_cconj_l1(space=self.domain)
+        elif self.pointwise_norm.exponent == 2:
+            return proximal_cconj_l1(space=self.domain, isotropic=True)
+        else:
+            raise NotImplementedError('`proximal` only implemented for p = 1 '
+                                      'or 2')
+
+    @property
+    def convex_conj(self):
+        """The convex conjugate is the the group L1-norm."""
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: Not sure I like a docstring that says "The X is Y." For me that would be the second line in the docstring.

File: odl/test/largescale/solvers/nonsmooth/proximal_operator_slow_test.py [hunk 9]
```diff
@@ -58,15 +53,16 @@ def offset(request):
 
 
 dual_params = [False, True]
-dual_ids = [' offset = {} '.format(str(p).ljust(5)) for p in offset_params]
+dual_ids = [' dual = {} '.format(str(p).ljust(5)) for p in offset_params]
```
Comment (kohr-h) [inline odl/test/largescale/solvers/nonsmooth/proximal_operator_slow_test.py]: dual_params

File: odl/test/largescale/solvers/nonsmooth/proximal_operator_slow_test.py (hunk 4 above)
Comment (kohr-h) [inline odl/test/largescale/solvers/nonsmooth/proximal_operator_slow_test.py]: Make a new issue for it.

File: odl/solvers/functional/default_functionals.py [hunk 10]
```diff
@@ -124,6 +129,248 @@ def convex_conj(self):
...
+                                 linear=False)
+
+            def _call(self, x):
+                """Return ``self(x)``."""
+
+                p = functional.pointwise_norm.exponent
+
+                if functional.pointwise_norm.exponent == 1:
+                    result = np.abs(x)
+                    np.divide(x, result, out=result, where=result != 0)
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: It's extremely slow, thats the main problem

Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: Okay, not so cool.

**Pull request merged**